src/detect-krb5-name.c

```diff
@@ -0,0 +1,411 @@
+                                   uint8_t, void *, void *, const Signature *,
+                                   const SigMatchCtx *);
+static int DetectKrb5SNameSetup (DetectEngineCtx *, Signature *, const char *);
+static void DetectKrb5SNameRegisterTests (void);
+
+static void DetectKrb5NameFree (void *);
+
+static int DetectEngineInspectKRB5Generic(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+static int g_krb5_name_list_id = 0;
+
+/**
+ * \brief Registration function for krb5.cname and krb5.sname: keywords
+ *
+ * This function is called once in the 'lifetime' of the engine.
+ */
+void DetectKrb5NameRegister(void) {
+    sigmatch_table[DETECT_AL_KRB5_CNAME].name = "krb5.cname";
+    sigmatch_table[DETECT_AL_KRB5_CNAME].desc = "match Kerberos 5 ticket client name";
+    sigmatch_table[DETECT_AL_KRB5_CNAME].url = DOC_URL DOC_VERSION "/rules/kerberos-keywords.html#krb5-cname";
+    sigmatch_table[DETECT_AL_KRB5_CNAME].Match = NULL;
+    sigmatch_table[DETECT_AL_KRB5_CNAME].AppLayerTxMatch = DetectKrb5CNameMatch;
+    sigmatch_table[DETECT_AL_KRB5_CNAME].Setup = DetectKrb5CNameSetup;
+    sigmatch_table[DETECT_AL_KRB5_CNAME].Free = DetectKrb5NameFree;
+    sigmatch_table[DETECT_AL_KRB5_CNAME].RegisterTests = DetectKrb5CNameRegisterTests;
+
+    sigmatch_table[DETECT_AL_KRB5_SNAME].name = "krb5.sname";
+    sigmatch_table[DETECT_AL_KRB5_SNAME].desc = "match Kerberos 5 ticket server name";
+    sigmatch_table[DETECT_AL_KRB5_SNAME].url = DOC_URL DOC_VERSION "/rules/kerberos-keywords.html#krb5-sname";
+    sigmatch_table[DETECT_AL_KRB5_SNAME].Match = NULL;
+    sigmatch_table[DETECT_AL_KRB5_SNAME].AppLayerTxMatch = DetectKrb5SNameMatch;
+    sigmatch_table[DETECT_AL_KRB5_SNAME].Setup = DetectKrb5SNameSetup;
+    sigmatch_table[DETECT_AL_KRB5_SNAME].Free = DetectKrb5NameFree;
+    sigmatch_table[DETECT_AL_KRB5_SNAME].RegisterTests = DetectKrb5SNameRegisterTests;
+
+    DetectAppLayerInspectEngineRegister("krb5_name",
+            ALPROTO_KRB5, SIG_FLAG_TOSERVER, 0,
+            DetectEngineInspectKRB5Generic);
+
+    DetectAppLayerInspectEngineRegister("krb5_name",
+            ALPROTO_KRB5, SIG_FLAG_TOCLIENT, 0,
+            DetectEngineInspectKRB5Generic);
+
+    /* set up the PCRE for keyword parsing */
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex, &parse_regex_study);
+
+    g_krb5_name_list_id = DetectBufferTypeRegister("krb5_name");
+    SCLogDebug("g_krb5_name_list_id %d", g_krb5_name_list_id);
+}
+
+static int DetectEngineInspectKRB5Generic(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
```

Comment:
This is a pattern matching one, right? I think this should be implemented differently then. Needs to be a 'sticky buffer' keyword, see smb_share for example.

Reply:
Hi,
after running some tests I'm not sure a sticky buffer is the best way to go: the client and server names are not simple strings, but PrincipalNames, which are encoded using a name-type and a list of strings (see https://tools.ietf.org/html/rfc4120#section-6.2). Using sticky buffers would mean that we can only match the entire buffer ?
Multiple strings are common, for ex in Active Directory, where you can find PrincipalNames like ( "ldap", "DOMAIN.COM" ), or ("ldap") only, quite often: using a keyword means we can match "ldap" easily. A substring match would not be great either, in some types the name can be encoded.
The current (proposed) approach leaves the matching on the rust side, to deal with these cases (it checks all substrings for a match).
Any idea ? Do you think a sticky buffer is better ?

Reply:
I'm not sure I understand the issue. If you need to match multiple things it can be done like with DNS (dns_query) where we can have multiple query buffers.

Can you mock up some example rules that would show the issue?

Reply:
The server name is a variable list of items (which are usually strings, but could be something else). For ex, it can be just ["krbtgt"], or ["krbtgt","DOMAIN.COM"], or it could contain more strings. I would like a rule `krb5_sname:krbtgt` to match any of these server names.

Another case is a Kerberos login using a certificate. I think the type used for the client name will be a X.509 DN, which is not a string but a DER object. In that case, `krb5_cname:toto` should match "CN=toto,DC=DOMAIN,DC=COM".

- Can I allocate multiple sticky buffers for one transaction, the number being known only after the transaction is created ?
- Where do I free these buffers ?

Also, `dns_query` seems to handle multiple buffers, but the code looks quite complicated. As I do not know what method is better (sticky buffers vs AppLayerTxMatch), I'm just asking for some guidance here on which one to choose :)

Reply:
I think there can be multiple ways: one would be with how the buffer is presented to the detection engine. Maybe it's possible to simply present it as a whole.

The dns_query approach seems best still, even if it's a bit more complex. It will perform much much better than any AppLayerTxMatch based approach.

Another thing to keep in mind is what we now have a transforms api. So perhaps a new transform could be used to take an input buffer and turn it into what you want.

The inspection buffer memory is managed for you, so no need to free things.